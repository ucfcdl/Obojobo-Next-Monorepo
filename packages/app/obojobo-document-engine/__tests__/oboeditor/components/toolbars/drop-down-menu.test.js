import { mount, shallow } from 'enzyme'
import React from 'react'

import DropDownMenu from '../../../../src/scripts/oboeditor/components/toolbars/drop-down-menu'

jest.useFakeTimers()

describe('Drop Down Menu', () => {
	test('DropDownMenu node', () => {
		const menu = [
			{ name: 'Undo', type: 'action', action: jest.fn() },
			{ name: 'Redo', type: 'action', action: jest.fn() }
		]
		const component = shallow(<DropDownMenu name="MockMenu" menu={menu} />)
		const tree = component.html()
		expect(tree).toMatchSnapshot()
	})

	test('DropDownMenu node with toggle-action', () => {
		const menu = [
			{ name: 'Show Placeholders', type: 'toggle-action', action: jest.fn(), value: true },
			{ name: 'Show Placeholders', type: 'toggle-action', action: jest.fn(), value: false }
		]
		const component = shallow(<DropDownMenu name="MockMenu" menu={menu} />)
		const tree = component.html()
		expect(tree).toMatchSnapshot()
	})

	test('DropDownMenu node with sub-menu', () => {
		const menu = [
			{ name: 'Undo', type: 'action', action: jest.fn() },
			{
				name: 'subMenu',
				type: 'sub-menu',
				menu: [{ name: 'SubItem', type: 'action', action: jest.fn() }]
			}
		]
		const component = mount(<DropDownMenu name="MockMenu" menu={menu} />)
		const tree = component.html()
		expect(tree).toMatchSnapshot()
	})

	test('DropDownMenu node opens and closes', () => {
		const menu = [
			{ name: 'Undo', type: 'action', action: jest.fn() },
			{ name: 'Redo', type: 'action', action: jest.fn() }
		]
		const component = mount(<DropDownMenu name="MockMenu" menu={menu} />)

		component
			.find('button')
			.at(0)
			.simulate('click')

		expect(component.html()).toMatchSnapshot()

		component
			.find('button')
			.at(0)
			.simulate('click')

		expect(component.html()).toMatchSnapshot()
	})

	test('DropDownMenu component opens and closes menu with keys', () => {
		const menu = [
			{ name: 'Undo', type: 'action', action: jest.fn() },
			{ name: 'Redo', type: 'action', action: jest.fn() }
		]
		const component = mount(<DropDownMenu name="MockMenu" menu={menu} />)

		component
			.find('div')
			.at(0)
			.simulate('keyDown', {
				key: 'ArrowRight',
				stopPropagation: jest.fn()
			})

		expect(component.html()).toMatchSnapshot()

		component
			.find('div')
			.at(0)
			.simulate('keyDown', {
				key: 'ArrowLeft',
				stopPropagation: jest.fn()
			})

		expect(component.html()).toMatchSnapshot()
	})

	test('DropDownMenu component moves up and down with keys', () => {
		const menu = [
			{ name: 'Undo', type: 'action', action: jest.fn() },
			{ name: 'Redo', type: 'action', action: jest.fn() }
		]
		const component = mount(<DropDownMenu name="MockMenu" menu={menu} />)

		component
			.find('div')
			.at(0)
			.simulate('keyDown', {
				key: 'ArrowUp',
				stopPropagation: jest.fn()
			})

		expect(component.state()).toMatchSnapshot()

		component
			.find('div')
			.at(0)
			.simulate('keyDown', {
				key: 'ArrowDown',
				stopPropagation: jest.fn()
			})

		expect(component.state()).toMatchSnapshot()
	})

	test('DropDownMenu component closes menu when unfocused', () => {
		const menu = [
			{ name: 'Undo', type: 'action', action: jest.fn() },
			{ name: 'Redo', type: 'action', action: jest.fn() }
		]
		const component = mount(<DropDownMenu name="MockMenu" menu={menu} />)

		const html = component
			.find('div')
			.at(0)
			.simulate('blur')
			.html()

		jest.runAllTimers()

		expect(html).toMatchSnapshot()
	})

	test('DropDownMenu component cancels menu closure when focused', () => {
		const menu = [
			{ name: 'Undo', type: 'action', action: jest.fn() },
			{ name: 'Redo', type: 'action', action: jest.fn() }
		]
		const component = mount(<DropDownMenu name="MockMenu" menu={menu} />)

		const html = component
			.find('div')
			.at(0)
			.simulate('focus')
			.html()

		jest.runAllTimers()

		expect(html).toMatchSnapshot()
	})
})
