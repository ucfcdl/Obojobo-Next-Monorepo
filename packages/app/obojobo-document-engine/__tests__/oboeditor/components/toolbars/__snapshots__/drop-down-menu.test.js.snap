// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drop Down Menu DropDownMenu component cancels menu closure when focused 1`] = `"<div class=\\"dropdown  is-not-open\\" tabindex=\\"-1\\"><button class=\\"menu-title\\">MockMenu</button><div class=\\"menu-items\\"><button>Undo</button><button>Redo</button></div></div>"`;

exports[`Drop Down Menu DropDownMenu component closes menu when unfocused 1`] = `"<div class=\\"dropdown  is-not-open\\" tabindex=\\"-1\\"><button class=\\"menu-title\\">MockMenu</button><div class=\\"menu-items\\"><button>Undo</button><button>Redo</button></div></div>"`;

exports[`Drop Down Menu DropDownMenu component moves up and down with keys 1`] = `
Object {
  "currentFocus": 1,
  "isFocused": false,
  "isOpen": false,
}
`;

exports[`Drop Down Menu DropDownMenu component moves up and down with keys 2`] = `
Object {
  "currentFocus": 0,
  "isFocused": false,
  "isOpen": false,
}
`;

exports[`Drop Down Menu DropDownMenu component opens and closes menu with keys 1`] = `"<div class=\\"dropdown  is-open\\" tabindex=\\"-1\\"><button class=\\"menu-title\\">MockMenu</button><div class=\\"menu-items\\"><button>Undo</button><button>Redo</button></div></div>"`;

exports[`Drop Down Menu DropDownMenu component opens and closes menu with keys 2`] = `"<div class=\\"dropdown  is-not-open\\" tabindex=\\"-1\\"><button class=\\"menu-title\\">MockMenu</button><div class=\\"menu-items\\"><button>Undo</button><button>Redo</button></div></div>"`;

exports[`Drop Down Menu DropDownMenu node 1`] = `"<div class=\\"dropdown  is-not-open\\" tabindex=\\"-1\\"><button class=\\"menu-title\\">MockMenu</button><div class=\\"menu-items\\"><button>Undo</button><button>Redo</button></div></div>"`;

exports[`Drop Down Menu DropDownMenu node opens and closes 1`] = `"<div class=\\"dropdown  is-open\\" tabindex=\\"-1\\"><button class=\\"menu-title\\">MockMenu</button><div class=\\"menu-items\\"><button>Undo</button><button>Redo</button></div></div>"`;

exports[`Drop Down Menu DropDownMenu node opens and closes 2`] = `"<div class=\\"dropdown  is-not-open\\" tabindex=\\"-1\\"><button class=\\"menu-title\\">MockMenu</button><div class=\\"menu-items\\"><button>Undo</button><button>Redo</button></div></div>"`;

exports[`Drop Down Menu DropDownMenu node with sub-menu 1`] = `"<div class=\\"dropdown  is-not-open\\" tabindex=\\"-1\\"><button class=\\"menu-title\\">MockMenu</button><div class=\\"menu-items\\"><button>Undo</button><div class=\\"dropdown  is-not-open\\" tabindex=\\"-1\\"><button class=\\"menu-title\\">subMenu</button><div class=\\"menu-items\\"><button>SubItem</button></div></div></div></div>"`;

exports[`Drop Down Menu DropDownMenu node with toggle-action 1`] = `"<div class=\\"dropdown  is-not-open\\" tabindex=\\"-1\\"><button class=\\"menu-title\\">MockMenu</button><div class=\\"menu-items\\"><button>Show Placeholders<span>âœ”</span></button><button>Show Placeholders</button></div></div>"`;
